src/librustdoc/doctest.rs

```diff
@@ -749,28 +755,27 @@ impl Tester for Collector {
             _ => PathBuf::from(r"doctest.rs"),
         };
 
+        // For example `module/file.rs` would become `module_file_rs`
+        let test_id = filename
+            .to_string()
+            .chars()
+            .map(|c| if c.is_ascii_alphanumeric() { c } else { '_' })
+            .collect::<String>();
+        let test_id = format!(
+            "{file}_{line}_{number}",
+            file = test_id,
+            line = line,
+            number = {
+                // Increases the current test number, if this file already
+                // exists or it creates a new entry with a test number of 0.
+                self.visited_tests
+                    .entry((test_id.clone(), line))
+                    .and_modify(|v| *v += 1)
+                    .or_insert(0)
+            },
+        );
         let outdir = if let Some(mut path) = options.persist_doctests.clone() {
-            // For example `module/file.rs` would become `module_file_rs`
-            let folder_name = filename
-                .to_string()
-                .chars()
-                .map(|c| if c == '\\' || c == '/' || c == '.' { '_' } else { c })
```

Comment:
I don't know the background here. Some of the changes you are proposing seem to have significant side effects that might affect downstream testing, especially if not limited to "coverage"-enabled tests.

1. Is it OK that `cratename` is no longer part of the `path`?

2. Why change the path at all?

3. And I guess I understand the need to change the `map()` for the `test_id` to check `is_ascii_alphanumeric` since you need to use it as a function name identifier, but again, using that in the path name as well seems like a shortcut with small benefit but a downstream effect that is broader in scope than just fixing doctest coverage.

If other compiler team members are OK with this, that's fine with me, but I want to make sure someone else takes a look.

Reply:
Maybe @ollie27 would be a good reviewer? I'm not very familiar with this part of rustdoc.

Reply:
Yes, the change is mostly to be able to use it as part of the identifier in code. The filename is only used for `--persist-doctests`, which is basically generating kind-of "random" filenames for each test. I think most tools using this will just enumerate all the executables in the `--persist-doctests` directory. Also, I recently fixed that part of the code already because it was broken on windows (did not replace `\` properly)